src/semidiscretization/semidiscretization_coupled.jl

```diff
@@ -558,28 +559,15 @@ function copy_to_coupled_boundary!(boundary_conditions::Union{Tuple, NamedTuple}
                               boundary_conditions...)
 end
 
-function mesh_equations_solver_cache(other_semi_index, i, semi_, semi_tuple...)
```

Comment:
We made this change to avoid excessive memory allocations/avoid type instability. Have you check that with your chances the memory allocations stay limited?

Reply:
As I said, this caused the memory allocations due to type instabilities. The type of the return value depends on `i`, so this is not type-stable. I moved the type instability one level up, so that it doesn't happen in performance-relevant code, which still causes allocations, but fewer of them. 

You could also pass `Val(i)` instead of `i`, which would make this function here type-stable. But you'd still end up with a type instability when you call `Val(i)`, as the return value of `Val(i)` is also not statically known.